### ISSUE TITLE
[New rule] `avoid_collection_methods_with_unrelated_types` (Willing to PR)

### ISSUE THREAD

**Issue by fzyzcjy** [dart-code-checker/dart-code-metrics#701]
**Please describe what the rule should do:**
Please see https://github.com/dart-lang/linter/issues/1307

**If your rule is inspired by other please provide link to it:**
https://github.com/dart-lang/linter/issues/1307
https://stackoverflow.com/questions/68251477/dart-flutter-linter-rule-the-type-to-index-a-map-should-be-the-key-type-of-map

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

```dart
https://github.com/dart-lang/linter/issues/1307
```

**Are you willing to submit a pull request to implement this rule?**
Yes

**Comment by incendial** [comment]
@fzyzcjy thanks for opening the issue, we thought about implementing this rule some months ago and it looks doable. really happy to see that you want to contribute, it you have any questions, please let me know!

**Comment by fzyzcjy** [comment]
😄

**Comment by fzyzcjy** [comment]
@incendial 

> if you have any questions, please let me know!

I do have a question :) Could you please give some hints about implementing this? I am not an expert in dart linters

**Comment by fzyzcjy** [comment]
@incendial Already implemented. https://github.com/dart-code-checker/dart-code-metrics/pull/705

**Comment by incendial** [comment]
I'll take a look in 2-3 hours.

**Comment by incendial** [comment]
Reopening until this rule is not released in pub

**Comment by dkrutskikh** [comment]
Available in 4.12.0 release 🚀
